mono/metadata/threads.c

```diff
@@ -5984,29 +6014,120 @@ mono_threads_summarize (MonoContext *ctx, gchar **out, MonoStackHash *hashes)
 				sleep (1);
 				mono_memory_barrier ();
-				/*mono_threads_pthread_kill (info, SIGTERM);*/
-
-				// Pause this handler so other handlers can run
-				/*int signum;*/
-				MOSTLY_ASYNC_SAFE_PRINTF("Waiting for signalled thread to collect stacktrace\n");
-				/*sigsuspend (&sigset);*/
-				/*mono_threads_pthread_kill (info, SIGTERM);*/
-				/*sigprocmask (SIG_UNBLOCK, &old_sigset, NULL);*/
-				/*g_assert (success == 0);*/
+				const char *name = thread_summary_state_to_str (summarizing_thread_state);
+				MOSTLY_ASYNC_SAFE_PRINTF("Waiting for signalled thread %zu to collect stacktrace. Status: %s\n", tid, name);
+				count--;
+			}
+			if (count == 0) {
+				MonoBoolean timeout_abort_success;
+
+				// After timeout
+				// Thread may have been in lock or something, may have died
+				// Either way, didn't respond. Rather than failing, we have to
+				// skip it.
+				mono_memory_barrier ();
+				switch (summarizing_thread_state) {
+				case MONO_SUMMARY_MUTATE_SHARED:
+					g_error ("Timed out when writing json log!");
+					break;
+
+				case MONO_SUMMARY_EMPTY: {
+					g_assert (summarizing_thread == (MonoNativeThreadId) NULL);
+					break;
+
+				case MONO_SUMMARY_EXAMINE:
+					MOSTLY_ASYNC_SAFE_PRINTF("Timed out, thread did not finish dumping\n");
+
+					MonoNativeThreadId old_val = ves_icall_System_Threading_Interlocked_CompareExchange_IntPtr ((gpointer) &summarizing_thread, GINT_TO_POINTER(tid) /* set */, GINT_TO_POINTER(NULL) /* compare */);
+					if (old_val != GINT_TO_POINTER (NULL))
+						g_error ("Attempting to abandon dumping of thread, and thread changed.");
+
+					ves_icall_System_Threading_Interlocked_CompareExchange_Int_Success (&summarizing_thread_state, MONO_SUMMARY_EMPTY /* set */, MONO_SUMMARY_EXAMINE /* compare */, &timeout_abort_success);
+					if (!timeout_abort_success)
+						g_error ("Thread state changed during timeout abort");
+
+					break;
+				}
+				case MONO_SUMMARY_EXPECT: {
+					MOSTLY_ASYNC_SAFE_PRINTF("Timed out, thread did not respond to signal\n");
+
+					MonoNativeThreadId old_val = ves_icall_System_Threading_Interlocked_CompareExchange_IntPtr ((gpointer) &summarizing_thread, GINT_TO_POINTER(NULL) /* set */, GINT_TO_POINTER(tid) /* compare */);
+					if (tid != old_val)
+						g_error ("Attempting to abandon dumping of thread %zu, and thread changed to %zu.", tid, old_val);
+
+					ves_icall_System_Threading_Interlocked_CompareExchange_Int_Success (&summarizing_thread_state, MONO_SUMMARY_EMPTY /* set */, MONO_SUMMARY_EXPECT /* compare */, &timeout_abort_success);
+					if (!timeout_abort_success)
+						g_error ("Thread state changed during timeout abort");
+
+					break;
+				}
+				default:
+					g_assert_not_reached ();
+				}
 			}
 		}
+
+		// After dumping all the other threads, we dump our own.
+		summarizing_thread = current;
+	}
+
+	mono_memory_barrier ();
+
+	MOSTLY_ASYNC_SAFE_PRINTF("Self-reporting for thread %zu\n", current);
+	g_assert (current == summarizing_thread);
```

Comment:
You could be more verbose here and print the values of `current` and `summarizing_thread`.